modules/file-transfer/src/main/java/org/apache/ignite/internal/network/file/FileTransferServiceImpl.java

```diff
@@ -0,0 +1,402 @@
+            Path tempDirectory,
+            FileSender fileSender,
+            FileReceiver fileReceiver
+    ) {
+        this.topologyService = topologyService;
+        this.messagingService = messagingService;
+        this.tempDirectory = tempDirectory;
+        this.fileSender = fileSender;
+        this.fileReceiver = fileReceiver;
+    }
+
+    @Override
+    public void start() {
+        topologyService.addEventHandler(new TopologyEventHandler() {
+            @Override
+            public void onDisappeared(ClusterNode member) {
+                fileReceiver.cancelTransfersFromSender(member.id());
+            }
+        });
+
+        messagingService.addMessageHandler(FileTransferMessageType.class,
+                (message, senderConsistentId, correlationId) -> {
+                    if (message instanceof FileDownloadRequest) {
+                        processDownloadRequest((FileDownloadRequest) message, senderConsistentId, correlationId);
+                    } else if (message instanceof FileUploadRequest) {
+                        processUploadRequest((FileUploadRequest) message, senderConsistentId, correlationId);
+                    } else if (message instanceof FileTransferErrorMessage) {
+                        processFileTransferErrorMessage((FileTransferErrorMessage) message);
+                    } else if (message instanceof FileTransferInfoMessage) {
+                        processFileTransferInfo((FileTransferInfoMessage) message);
+                    } else if (message instanceof FileHeaderMessage) {
+                        processFileHeader((FileHeaderMessage) message);
+                    } else if (message instanceof FileChunkMessage) {
+                        processFileChunk((FileChunkMessage) message);
+                    } else {
+                        LOG.error("Unexpected message received: {}", message);
+                    }
+                });
+    }
+
+    private void processFileTransferErrorMessage(FileTransferErrorMessage message) {
+        fileReceiver.receiveFileTransferErrorMessage(message);
+    }
+
+    private void processUploadRequest(FileUploadRequest message, String senderConsistentId, long correlationId) {
+        completedFuture(UUID.randomUUID())
+                .thenCompose(transferId -> {
+                    Path directory = createTransferDirectory(transferId);
+                    FileTransferMessagesHandler handler = fileReceiver.registerTransfer(
+                            senderConsistentId,
+                            transferId,
+                            directory
+                    );
+                    FileUploadResponse response = factory.fileUploadResponse()
+                            .transferId(transferId)
+                            .build();
+                    Metadata metadata = message.metadata();
+                    return messagingService.respond(senderConsistentId, FILE_TRANSFERRING_CHANNEL, response, correlationId)
```

Comment:
Threading model here is not clear to me. All "compose" are not "async", how do you guarantee that network thread won't be executing your disk IO operations?

Reply:
I don't see how this clarifies my question. You're doing "messagingService.respond(...).thenCompose(...)", how is it related to executors in FileSender/FileReceiver? Do we guarantee that "handler.result()" will never be completed in that case, so that "handleUpload" is in its proper pool? I'd love to see that in a comment